src/Okta.Sdk/DefaultRetryStrategy.cs

```diff
@@ -0,0 +1,158 @@
+        {
+            if (requestTimeout > 0 && backoffSecondsDelta > requestTimeout)
+            {
+                throw new ArgumentException("The backoff delta cannot be greater than the request timeout");
+            }
+
+            _maxRetries = maxRetries;
+            _requestTimeout = requestTimeout;
+            _backoffSecondsDelta = backoffSecondsDelta;
+        }
+
+        /// <inheritdoc/>
+        public async Task<HttpResponseMessage> WaitAndRetryAsync(HttpRequestMessage request, CancellationToken cancellationToken, Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> operation)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (operation == null)
+            {
+                throw new ArgumentNullException(nameof(operation));
+            }
+
+            var numberOfRetries = 0;
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            do
+            {
+                var response = await operation(request, cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    if (IsRetryable(response) && numberOfRetries < _maxRetries &&
+                        (_requestTimeout <= 0 || stopwatch.Elapsed.Seconds < _requestTimeout))
+                    {
+                        numberOfRetries++;
+                        await CreateDelayAsync(response, cancellationToken).ConfigureAwait(false);
+                        response.Headers.TryGetValues("X-Okta-Request-Id", out var requestId);
+                        AddRetryOktaHeaders(request, requestId.FirstOrDefault(), numberOfRetries);
+                    }
+                    else
+                    {
+                        return response;
+                    }
+                }
+                catch (InvalidOperationException)
+                {
+                    return response;
+                }
+            }
+            while (true);
+        }
+
+        /// <summary>
+        /// Checks if a http response message should be retried
+        /// </summary>
+        /// <param name="response">The http response message</param>
+        /// <returns>True if the value is must be retried, otherwise false.</returns>
+        public virtual bool IsRetryable(HttpResponseMessage response)
+        {
+            var isRetryable = false;
+
+            if (response != null && _retryableStatusCodes.Contains(response.StatusCode))
+            {
+                isRetryable = true;
+            }
+
+            return isRetryable;
+        }
+
+        private void AddRetryOktaHeaders(HttpRequestMessage request, string requestId, int numberOfRetries)
+        {
+            if (!request.Headers.Contains("X-Okta-Retry-For"))
+            {
+                request.Headers.Add("X-Okta-Retry-For", requestId);
```

Comment:
nit: If `requestId` is `null` here what happens? Hopefully we don't get an exception or print an empty `X-Okta-Retry-For` header in the request.